docs/tech/rfc/0019-240310-cesium-race-conditions.md

```diff
@@ -7,129 +7,616 @@
-
-Lastly, we will discuss streamers: fundamentally, streamers do not make modifications to the database and can either be used to stream data being written to it or read from it. We can see that it is okay to have a streamer open on an inexistent unary database, as there will be no write or read requests made to it, so the streamer will receive nothing in its inlet. Eventually, this inlet will be closed.
-
-To recap, we can delete a channel if there are no open iterators or writers on it, and if it is not an index channel for other channels in the database.
+Concurrency is a double-edged sword: while it could make Cesium blazing-fast, it introduces
+additional complexity through parallelism – when many subprocesses run at the same time,
+they create conflicts, or race conditions, when they attempt to access common resources.
+As Cesium looks to become production-ready, it is crucial that Cesium must support
+Read, Write, and Delete (through a tombstone/garbage-collection system) operations
+at high speeds without them conflicting each other. One can imagine the possible
+consequences for unhandled race conditions in extreme scenarios such as launch control systems:
+it would not be desirable.
+
+# 4- Analysis
+
+## 4.0 Primer on the structure of Cesium
+
+At the highest level of abstraction, Cesium provides one unified interface to read, write,
+delete, and stream data from multiple channels.
+A Cesium DB comprises many channels, each one of which corresponding to a virtualDB
+or an unaryDB. A virtualDB contains no data, so it will not be a topic of our discussion.
+
+A unaryDB is a second level of abstraction in the database structure: it is composed of
+a domainDB and an index: the index is responsible for resolving locations of data, whereas
+the domainDB is responsible for storing it.
+
+A domainDB is the lowest level of abstraction: it provides means to write and read data from
+the file system, and maintains the index (see concepts).
+
+Each layer has their own entities and methods of reads, writes, deletes, and garbage collection;
+all interactions are passed down to the domain level to interact with the index and file system.
+
+## 4.1 Top-down view of race conditions in Cesium
+
+Protecting the integrity of data in Cesium is crucial.
+The three shared resources that we cannot allow concurrent access to are data stored on the file system,
+the index stored in memory, and its persisted counterpart on the file system.
+We will spend most of our time talking about the first two,
+since the persisted index is never read from until the database is closed and reopened.
+
+Each one of the read, write, delete, and garbage collection operations is a combination
+of accesses to the index and telemetry data. In the following diagram, we consider the
+life cycle of each of these operations and which resources they need to access.
+
+<div align="center">
+    <img src="img/0019-240529-cesium-race-conditions/Operation-Entities.png" width="90%" />
```

Comment:
I'm a little confused on the difference between "find pointer position" and "read index for pointer info". Isn't the position of the pointer stored in the index?